src/decay.rs

```diff
@@ -0,0 +1,123 @@
+use super::*;
+
+/// Computes `1 - e^(-x)` with numerical stability.
+/// Returns 0.0 at x=0, saturates to 1.0 as x increases.
+/// Used for EMA warmup bias correction.
+fn exponential_saturation(x: f64) -> f64 {
+    // Maximum ratio where `1 - e^(-x)` is distinguishable from 1.0.
+    // Beyond this, `e^(-x) < f64::EPSILON` and the subtraction rounds to exactly 1.0.
+    // Derived from: `-ln(f64::EPSILON) = 36.04`
+    -(-x.min(36.0)).exp_m1()
+}
+
+/// Calculates time bias based on how much history we have.
+/// Returns a value approaching 1.0 as elapsed time exceeds the window.
+pub(crate) fn calculate_time_bias(elapsed: Duration, window: Duration) -> f64 {
+    exponential_saturation(elapsed.as_secs_f64() / window.as_secs_f64())
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct DecayingAverage {
+    value: f64,
+    window: Duration,
+    last_update: Instant,
+}
+
+impl DecayingAverage {
+    pub(crate) fn new(window: Duration) -> Self {
+        Self {
+            value: 0.0,
+            window,
+            last_update: Instant::now(),
+        }
+    }
```

Comment:
Let's add a validator to this here constructor. The reason being I believe that creating a DecayingAverage with Duration::ZERO would cause division by zero and NaN propagation. We'll get a fail-fast at construction that will help us prevent subtle NaN bugs.
```suggestion
impl DecayingAverage {
    pub(crate) fn new(window: Duration) -> Self {
        assert!(
            !window.is_zero(),
            "DecayingAverage window must be non-zero"
        );
        Self {
            value: 0.0,
            window,
            last_update: Instant::now(),
        }
    }
}
```